File: static/solutions/cpp/search_list.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <memory>

#include "./Linked_list_prototype.h"

using std::make_shared;
using std::shared_ptr;

// @include
shared_ptr<ListNode<int>> SearchList(shared_ptr<ListNode<int>> L, int key) {
  while (L && L->data != key) {
    L = L->next;
  }
  // If key was not present in the list, L will have become null.
  return L;
}
// @exclude

int main(int argc, char* argv[]) {
  shared_ptr<ListNode<int>> L;
  L = make_shared<ListNode<int>>(ListNode<int>{
      2, make_shared<ListNode<int>>(ListNode<int>{
             4, make_shared<ListNode<int>>(ListNode<int>{3, nullptr})})});
  assert(L == SearchList(L, 2));
  assert(L->next == SearchList(L, 4));
  assert(L->next->next == SearchList(L, 3));
  assert(nullptr == SearchList(L, 100));
  return 0;
}

File: static/python/search_list.py
from linked_list_prototype import ListNode


# @include
def search_list(L, key):
    while L and L.data != key:
        L = L.next
    # If key was not present in the list, L will have become null.
    return L
# @exclude


def main():
    L = ListNode(2, ListNode(4, ListNode(3, None)))
    assert L is search_list(L, 2)
    assert L.next is search_list(L, 4)
    assert L.next.next is search_list(L, 3)
    assert not search_list(L, 100)


if __name__ == '__main__':
    main()
